internal/resources/cluster/data_source_cluster.go

```diff
@@ -25,116 +25,162 @@ import (
 
 func DataSourceTMCCluster() *schema.Resource {
 	return &schema.Resource{
-		ReadContext: dataSourceTMCClusterRead,
+		ReadContext: schema.ReadContextFunc(dataSourceTMCClusterRead(withTerraformConstructType(constructTypeDataSource))),
 		Schema:      clusterSchema,
 	}
 }
 
-func dataSourceTMCClusterRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
-	config := m.(authctx.TanzuContext)
-
-	// Warning or errors can be collected in a slice type
-	var (
-		diags diag.Diagnostics
-		resp  *clustermodel.VmwareTanzuManageV1alpha1ClusterGetClusterResponse
-		err   error
-	)
-
-	getClusterResourceRetryableFn := func() (retry bool, err error) {
-		resp, err = config.TMCConnection.ClusterResourceService.ManageV1alpha1ClusterResourceServiceGet(constructFullname(d))
-		if err != nil {
-			if clienterrors.IsNotFoundError(err) {
-				_ = schema.RemoveFromState(d, m)
-				return false, nil
-			}
+type terraformOption func(*terraformConfig)
 
-			return true, errors.Wrapf(err, "Unable to get Tanzu Mission Control cluster entry, name : %s", d.Get(NameKey))
-		}
+type (
+	terraformConstruct string
+	operationType      string
+)
 
-		d.SetId(resp.Cluster.Meta.UID)
+const (
```

Comment:
Why do we need so many states, wouldn't it complicate things?
I feel that passing these caller type in the context would be a good way to pass to these functions.
and the values are 
"DataRead, RefreshState, CreateState, UpdateState", most cases there will be only "DataRead & RefreshState"